XMOS VocalFusion I2S Driver Update

BohdanBuinich opened the pull request.

BohdanBuinich committed:
XMOS VocalFusion I2S Driver Update

j1nx commented:
Lol, looks like you did what I was trying to figure out.

will test it in the morning

BohdanBuinich commented:
> Lol, looks like you did what I was trying to figure out.
> 
> will test it in the morning

please wait, I broke something while cleaning the code , I'll inform when it works again

j1nx commented:
Looking good already. You also implemented some old todo's.

will await your go ahead

BohdanBuinich commented:
> Looking good already. You also implemented some old todo's.
> 
> will await your go ahead

Everything is ready.
Perhaps I made an overhead with determining the current RPI board, but I did not find another way out

BohdanBuinich commented:
But it still doesn't work, on RPI 5, the driver doesn't run probe at all.....

It probably won't work because they've changed the hardware a bit.

j1nx commented:
Will give it a test run today and report back.

Going forward for at least the rpi5, I will open an issue here at this repo so we can work together. The current driver is hacked together based on my own research from a while back. I will gather all those thoughts and referenced from back then and post them there. There are some thing we can and most likely need todo to modernize this driver and associated overlays.

Will get back to you.

j1nx commented:
It also fails for me on the rpi4. Below the dmesg;
```
[    7.624171] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/i2s@7e203000/status
[    7.624205] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/compatible
[    7.624212] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/i2s-controller
[    7.624219] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/status
[    7.624225] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clocks
[    7.624231] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-names
[    7.624237] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-frequency
[    7.624245] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-names
[    7.624253] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-0
[    7.641745] vocalfusion_soundcard: loading out-of-tree module taints kernel.
[    7.642670] mclk at gpio4 set to: 24576000 Hz
[    7.648779] request module load 'bcm2708-dmaengine': 0
[    7.649046] register platform device 'asoc-simple-card': 0
[    7.649063] Unable to handle kernel NULL pointer dereference at virtual address 0000000000000010
[    7.649066] Mem abort info:
[    7.649067]   ESR = 0x0000000096000005
[    7.649069]   EC = 0x25: DABT (current EL), IL = 32 bits
[    7.649072]   SET = 0, FnV = 0
[    7.649074]   EA = 0, S1PTW = 0
[    7.649076]   FSC = 0x05: level 1 translation fault
[    7.649078] Data abort info:
[    7.649079]   ISV = 0, ISS = 0x00000005, ISS2 = 0x00000000
[    7.649081]   CM = 0, WnR = 0, TnD = 0, TagAccess = 0
[    7.649084]   GCS = 0, Overlay = 0, DirtyBit = 0, Xs = 0
[    7.649087] user pgtable: 4k pages, 39-bit VAs, pgdp=00000000461e1000
[    7.649090] [0000000000000010] pgd=0000000000000000, p4d=0000000000000000, pud=0000000000000000
[    7.649099] Internal error: Oops: 0000000096000005 [#1] PREEMPT_RT SMP
[    7.649103] Modules linked in: vocalfusion_soundcard(O+) joydev spidev brcmfmac_wcc brcmfmac brcmutil v3d cfg80211 gpu_sched raspberrypi_ts raspberrypi_hwmon rfkill i2c_bcm2835 raspberrypi_gpiomem spi_bcm2835 bcm2835_isp(C) bcm2835_v4l2(C) snd_soc_bcm2835_i2s bcm2835_codec(C) bcm2835_mmal_vchiq(C) vc_sm_cma(C) snd_bcm2835(C) videobuf2_vmalloc rpivid_hevc(C) v4l2_mem2mem videobuf2_dma_contig videobuf2_memops videobuf2_v4l2 videodev vc4 videobuf2_common snd_soc_hdmi_codec drm_display_helper cec drm_dma_helper snd_soc_core snd_compress snd_pcm_dmaengine snd_pcm snd_timer snd rpi_backlight backlight nvmem_rmem vfat uio_pdrv_genirq uio sch_fq_codel i2c_dev fuse nfsd
[    7.649176] CPU: 1 PID: 378 Comm: systemd-udevd Tainted: G         C O       6.6.14-rt21-ovos-buildroot #1
[    7.649182] Hardware name: Raspberry Pi Foundation Raspberry Pi 4 Model B/Raspberry Pi 4 Model B, BIOS UEFI Firmware v1.33 03/07/2022
[    7.649185] pstate: 00000005 (nzcv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[    7.649189] pc : validate_desc+0x28/0xa0
[    7.649208] lr : gpiod_direction_output+0x30/0x140
[    7.649213] sp : ffffffc08066b840
[    7.649214] x29: ffffffc08066b840 x28: ffffffd40955e838 x27: 0000000000000001
[    7.649221] x26: ffffffc08066bc60 x25: ffffffc08066bc10 x24: ffffff80473f8380
[    7.649226] x23: ffffffd484668960 x22: 0000000000000000 x21: ffffffd40955e050
[    7.649231] x20: 0000000000000001 x19: 0000000000000010 x18: ffffffc08066b4a0
[    7.649236] x17: 4554535953425553 x16: ffffffd483e89060 x15: ffffffd4851b45ba
[    7.649241] x14: 0000000000000000 x13: 30203a2764726163 x12: 2d656c706d69732d
[    7.649248] x11: 636f736127206563 x10: 69766564206d726f x9 : ffffffd483e89090
[    7.649253] x8 : 00000000ffffefff x7 : ffffffd484fe8628 x6 : 0000000000000000
[    7.649258] x5 : ffffff807dfa5dc8 x4 : 0000000000000000 x3 : 0000000000000000
[    7.649263] x2 : ffffffd48463e4a0 x1 : ffffffd48463e6f0 x0 : 0000000000000010
[    7.649269] Call trace:
[    7.649273]  validate_desc+0x28/0xa0
[    7.649279]  gpiod_direction_output+0x30/0x140
[    7.649284]  vocalfusion_soundcard_probe+0xe8/0x170 [vocalfusion_soundcard]
[    7.649301]  platform_probe+0x70/0xd0
[    7.649308]  really_probe+0x150/0x2c8
[    7.649316]  __driver_probe_device+0x80/0x140
[    7.649322]  driver_probe_device+0xdc/0x178
[    7.649327]  __driver_attach+0x9c/0x1b8
[    7.649331]  bus_for_each_dev+0x80/0xe8
[    7.649336]  driver_attach+0x2c/0x40
[    7.649341]  bus_add_driver+0xec/0x1f8
[    7.649345]  driver_register+0x68/0x138
[    7.649352]  __platform_driver_register+0x30/0x48
[    7.649355]  vocalfusion_soundcard_driver_init+0x28/0xff8 [vocalfusion_soundcard]
[    7.649364]  do_one_initcall+0x60/0x2c0
[    7.649369]  do_init_module+0x5c/0x218
[    7.649380]  load_module+0x1da8/0x2058
[    7.649386]  init_module_from_file+0x90/0xd8
[    7.649391]  __arm64_sys_finit_module+0x14c/0x2f0
[    7.649397]  invoke_syscall+0x50/0x128
[    7.649404]  el0_svc_common.constprop.0+0xc8/0xf0
[    7.649410]  do_el0_svc+0x24/0x38
[    7.649415]  el0_svc+0x40/0xe8
[    7.649422]  el0t_64_sync_handler+0x100/0x130
[    7.649427]  el0t_64_sync+0x190/0x198
[    7.649433] Code: b140041f f9000bf3 aa0003f3 54000168 (f9400003)
[    7.649436] ---[ end trace 0000000000000000 ]---
[    7.662437] snd_soc_simple_card: unknown parameter 'index' ignored
```

I will go through all my old notes and thoughts and will open the Issue tracker where we can discuss our approach and see if we can tackle this together in teh way is should be done, instead of based on the ancient myloader.ko driver

j1nx commented:
Output of gpioinfo does also reveal the reset and power GPIO's are not properly set to output.
```
ovos@OpenVoiceOS-840dca:~ $ gpioinfo
gpiochip0 - 58 lines:
        line   0:       "ID_SDA"                input
        line   1:       "ID_SCL"                input
        line   2:       "GPIO2"                 input
        line   3:       "GPIO3"                 input
        line   4:       "GPIO4"                 input
        line   5:       "GPIO5"                 input
        line   6:       "GPIO6"                 input
        line   7:       "GPIO7"                 output active-low consumer="spi0 CS1"
        line   8:       "GPIO8"                 output active-low consumer="spi0 CS0"
        line   9:       "GPIO9"                 input
        line  10:       "GPIO10"                input
        line  11:       "GPIO11"                input
        line  12:       "GPIO12"                input
        line  13:       "GPIO13"                input
        line  14:       "GPIO14"                input
        line  15:       "GPIO15"                input
        line  16:       "GPIO16"                input
        line  17:       "GPIO17"                input
        line  18:       "GPIO18"                input
        line  19:       "GPIO19"                input
        line  20:       "GPIO20"                input
        line  21:       "GPIO21"                input
        line  22:       "GPIO22"                input
        line  23:       "GPIO23"                input
        line  24:       "GPIO24"                input
        line  25:       "GPIO25"                input
        line  26:       "GPIO26"                input
        line  27:       "GPIO27"                input
        line  28:       "RGMII_MDIO"            input
        line  29:       "RGMIO_MDC"             input
        line  30:       "CTS0"                  input
        line  31:       "RTS0"                  output
        line  32:       "TXD0"                  input
        line  33:       "RXD0"                  input
        line  34:       "SD1_CLK"               input
        line  35:       "SD1_CMD"               input
        line  36:       "SD1_DATA0"             input
        line  37:       "SD1_DATA1"             input
        line  38:       "SD1_DATA2"             input
        line  39:       "SD1_DATA3"             input
        line  40:       "PWM0_MISO"             input
        line  41:       "PWM1_MOSI"             input
        line  42:       "STATUS_LED_G_CLK"      output consumer="ACT"
        line  43:       "SPIFLASH_CE_N"         input
        line  44:       "SDA0"                  input
        line  45:       "SCL0"                  input
        line  46:       "RGMII_RXCLK"           input
        line  47:       "RGMII_RXCTL"           input
        line  48:       "RGMII_RXD0"            input
        line  49:       "RGMII_RXD1"            input
        line  50:       "RGMII_RXD2"            input
        line  51:       "RGMII_RXD3"            input
        line  52:       "RGMII_TXCLK"           input
        line  53:       "RGMII_TXCTL"           input
        line  54:       "RGMII_TXD0"            input
        line  55:       "RGMII_TXD1"            input
        line  56:       "RGMII_TXD2"            input
        line  57:       "RGMII_TXD3"            input
gpiochip1 - 8 lines:
        line   0:       "BT_ON"                 output
        line   1:       "WL_ON"                 output
        line   2:       "PWR_LED_OFF"           output active-low consumer="PWR"
        line   3:       "GLOBAL_RESET"          output
        line   4:       "VDD_SD_IO_SEL"         output consumer="vdd-sd-io"
        line   5:       "CAM_GPIO"              output consumer="cam1_regulator"
        line   6:       "SD_PWR_ON"             output consumer="regulator-sd-vcc"
        line   7:       "SD_OC_N"               input
```
After flashing the firmware blob thos two GPIO pins need to be pulled to startup the whole XMOS chip with the right clk at GPIO4 at the right Mhz for syncing the 48kHz waves. (More on that later)

BohdanBuinich commented:
> [    7.649279]  gpiod_direction_output+0x30/0x140
> [    7.649284]  vocalfusion_soundcard_probe+0xe8/0x170 [vocalfusion_soundcard]

I removed call gpiod_direction_output func

and

> [    7.642670] mclk at gpio4 set to: 24576000 Hz
> [    7.648779] request module load 'bcm2708-dmaengine': 0

I don't see the new print which I added:
dev_info(dev, "Module initialized with platform: %s, DMA engine: %s\n", card_platform, dmaengine);

> [    7.642670] mclk at gpio4 set to: 24576000 Hz

and this print now look like: dev_info(dev, "mclk set to: %lu Hz\n", clk_get_rate(mclk));
without GPIO number

look like you still have old repo, either I'm missing something, I'm going to rebuild everything

BohdanBuinich commented:
what I see is:
```c
[    8.963361] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/i2s@7e203000/status
[    8.963380] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/compatible
[    8.963386] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/i2s-controller
[    8.963392] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/status
[    8.963398] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clocks
[    8.963403] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-names
[    8.963409] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-frequency
[    8.963414] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-names
[    8.963420] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-0
[    8.963425] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pwr-gpio
[    8.963430] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/rst-gpio
[    9.009560] vocalfusion_soundcard: loading out-of-tree module taints kernel.
[    9.012212] vocalfusion-driver soc:sound: Module initialized with platform: fe203000.i2s, DMA engine: bcm2708-dmaengine
[    9.012454] vocalfusion-driver soc:sound: mclk set to: 24576000 Hz
[    9.030627] snd_soc_simple_card: unknown parameter 'index' ignored
[    9.826353] bcmgenet fd580000.ethernet: configuring instance for external RGMII (RX delay)
[    9.827370] bcmgenet fd580000.ethernet eth0: Link is Down
[    9.880791] brcmfmac: brcmf_cfg80211_set_power_mgmt: power save enabled
[   10.026660] systemd[551]: memfd_create() called without MFD_EXEC or MFD_NOEXEC_SEAL set
[   12.539375] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/pwm@7e20c000/pinctrl-names
[   12.539393] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/pwm@7e20c000/pinctrl-0
[   12.539400] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/pwm@7e20c000/assigned-clock-rates
[   12.539406] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/pwm@7e20c000/status
[   12.539482] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp0
[   12.539489] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp0_hyst
[   12.539494] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp1
[   12.539500] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp1_hyst
[   12.539505] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp2
[   12.539509] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp2_hyst
[   12.539514] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp3
[   12.539519] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /__overrides__/poe_fan_temp3_hyst
[   12.541610] OF: /sj201-rev10-pwm-fan@0: #pwm-cells = 3 found 2
[   12.559340] OF: /sj201-rev10-pwm-fan@0: #pwm-cells = 3 found 2
[   12.559356] of_pwm_get(): can't parse "pwms" property
[   12.559361] pwm-fan sj201-rev10-pwm-fan@0: error -EINVAL: Could not get PWM
[   12.559366] pwm-fan: probe of sj201-rev10-pwm-fan@0 failed with error -22
[   12.642385] input: soc:sj201_buttons as /devices/platform/soc/soc:sj201_buttons/input/input1
[   12.893334] bcmgenet fd580000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
[   19.139413] systemd-journald[155]: Oldest entry in /run/log/journal/c4e4e6268855400bbc329f1e97200a6b/system.journal is older than the configured file retention duration (1month), suggesting rotation.
[   19.139434] systemd-journald[155]: /run/log/journal/c4e4e6268855400bbc329f1e97200a6b/system.journal: Journal header limits reached or header out-of-date, rotating.
ovos@OpenVoiceOS-0822a6:~ $ i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- 2c -- -- 2f
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --
```

@j1nx could you try to retest on my branch https://github.com/OpenVoiceOS/ovos-buildroot/pull/213

```
[   12.541610] OF: /sj201-rev10-pwm-fan@0: #pwm-cells = 3 found 2
[   12.559340] OF: /sj201-rev10-pwm-fan@0: #pwm-cells = 3 found 2
[   12.559356] of_pwm_get(): can't parse "pwms" property
[   12.559361] pwm-fan sj201-rev10-pwm-fan@0: error -EINVAL: Could not get PWM
[   12.559366] pwm-fan: probe of sj201-rev10-pwm-fan@0 failed with error -22
```

this one has not been fixed yet, perhaps we will address it in the next PR

BohdanBuinich commented:
I think our overlay might look something like this https://github.com/raspberrypi/linux/blob/rpi-6.6.y/arch/arm/boot/dts/overlays/ugreen-dabboard-overlay.dts

And in kernel code, we need to set the Clock and GPIOs

j1nx commented:
I guess I screwed up a bit with testing. Indeed it looks like I did not test the latest code.

I was distracted the whole weekend. I will create the Issue with the (old) information at hand, during the day,

j1nx commented:
Gathered my thoughts and old links here;
https://github.com/OpenVoiceOS/VocalFusionDriver/issues/4


Will re-do my testing with the current code base this evening.

j1nx commented:
Had a bit of busy days but never reported back the re-do results of my test run.

In short. This current code base does work on the Mark2 retail kit with teh Raspberry Pi4 - 2GB

```
[    7.797452] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/i2s@7e203000/status
[    7.797470] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/compatible
[    7.797476] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/i2s-controller
[    7.797482] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/status
[    7.797488] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clocks
[    7.797493] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-names
[    7.797499] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/clock-frequency
[    7.797504] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-names
[    7.797510] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pinctrl-0
[    7.797515] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/pwr-gpio
[    7.797521] OF: overlay: WARNING: memory leak will occur if overlay removed, property: /soc/sound/rst-gpio
[    7.807469] vocalfusion_soundcard: loading out-of-tree module taints kernel.
[    7.808706] vocalfusion-driver soc:sound: Module initialized with platform: fe203000.i2s, DMA engine: bcm2708-dmaengine
[    7.809641] vocalfusion-driver soc:sound: mclk set to: 24576000 Hz
[    7.824674] snd_soc_simple_card: unknown parameter 'index' ignored

ovos@OpenVoiceOS-840dca:~ $ journalctl -u i2csound
Dec 20 18:15:50 OpenVoiceOS systemd[1]: Starting I2C Sound configuration service...
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: SJ201LED detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: TAS5806 detection result: true
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: ADAFRUIT detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: AIYVOICEBONNET detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: WM8XXX detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: RESPEAKER6 detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: RESPEAKER4 detection result: false
Dec 20 18:15:50 OpenVoiceOS ovos-i2csound[419]: Installing and configuring SJ-201 HAT
Dec 20 18:15:51 OpenVoiceOS-840dca ovos-i2csound[473]: Read file "/usr/lib/firmware/xvf3510/app_xvf3510_int_spi_boot_v4_1_0.bin" size: 196608 Bytes
Dec 20 18:15:51 OpenVoiceOS-840dca ovos-i2csound[473]: Sending complete
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: Deep sleep   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: Deep sleep   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: Deep sleep   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: Deep sleep   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Clock fault (reg: 0x71)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: HIZ   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Clock fault (reg: 0x71)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: HIZ   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: FS_MON: FS Error   (reg: 0x37)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Clock fault (reg: 0x71)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[544]: Run Status: HIZ   (reg: 0x68)
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[419]: Assume revision-10 SJ-201 board
Dec 20 18:15:53 OpenVoiceOS-840dca ovos-i2csound[419]: Configuring buttons
Dec 20 18:15:53 OpenVoiceOS-840dca systemd[1]: i2csound.service: Deactivated successfully.
Dec 20 18:15:53 OpenVoiceOS-840dca systemd[1]: Finished I2C Sound configuration service.

ovos@OpenVoiceOS-840dca:~ $ gpioinfo
gpiochip0 - 58 lines:
        line   0:       "ID_SDA"                input
        line   1:       "ID_SCL"                input
        line   2:       "GPIO2"                 input
        line   3:       "GPIO3"                 input
        line   4:       "GPIO4"                 input
        line   5:       "GPIO5"                 input
        line   6:       "GPIO6"                 input
        line   7:       "GPIO7"                 output active-low consumer="spi0 CS1"
        line   8:       "GPIO8"                 output active-low consumer="spi0 CS0"
        line   9:       "GPIO9"                 input
        line  10:       "GPIO10"                input
        line  11:       "GPIO11"                input
        line  12:       "GPIO12"                input
        line  13:       "GPIO13"                input
        line  14:       "GPIO14"                input
        line  15:       "GPIO15"                input
        line  16:       "GPIO16"                output consumer="pwr"
        line  17:       "GPIO17"                input
        line  18:       "GPIO18"                input
        line  19:       "GPIO19"                input
        line  20:       "GPIO20"                input
        line  21:       "GPIO21"                input
        line  22:       "GPIO22"                input
        line  23:       "GPIO23"                input
        line  24:       "GPIO24"                input
        line  25:       "GPIO25"                input
        line  26:       "GPIO26"                input
        line  27:       "GPIO27"                output consumer="rst"
        line  28:       "RGMII_MDIO"            input
        line  29:       "RGMIO_MDC"             input
        line  30:       "CTS0"                  input
        line  31:       "RTS0"                  output
        line  32:       "TXD0"                  input
        line  33:       "RXD0"                  input
        line  34:       "SD1_CLK"               input
        line  35:       "SD1_CMD"               input
        line  36:       "SD1_DATA0"             input
        line  37:       "SD1_DATA1"             input
        line  38:       "SD1_DATA2"             input
        line  39:       "SD1_DATA3"             input
        line  40:       "PWM0_MISO"             input
        line  41:       "PWM1_MOSI"             input
        line  42:       "STATUS_LED_G_CLK"      output consumer="ACT"
        line  43:       "SPIFLASH_CE_N"         input
        line  44:       "SDA0"                  input
        line  45:       "SCL0"                  input
        line  46:       "RGMII_RXCLK"           input
        line  47:       "RGMII_RXCTL"           input
        line  48:       "RGMII_RXD0"            input
        line  49:       "RGMII_RXD1"            input
        line  50:       "RGMII_RXD2"            input
        line  51:       "RGMII_RXD3"            input
        line  52:       "RGMII_TXCLK"           input
        line  53:       "RGMII_TXCTL"           input
        line  54:       "RGMII_TXD0"            input
        line  55:       "RGMII_TXD1"            input
        line  56:       "RGMII_TXD2"            input
        line  57:       "RGMII_TXD3"            input
gpiochip1 - 8 lines:
        line   0:       "BT_ON"                 output
        line   1:       "WL_ON"                 output
        line   2:       "PWR_LED_OFF"           output active-low consumer="PWR"
        line   3:       "GLOBAL_RESET"          output
        line   4:       "VDD_SD_IO_SEL"         output consumer="vdd-sd-io"
        line   5:       "CAM_GPIO"              output consumer="cam1_regulator"
        line   6:       "SD_PWR_ON"             output consumer="regulator-sd-vcc"
        line   7:       "SD_OC_N"               input


ovos@OpenVoiceOS-840dca:~ $ aplay -lL
null
    Discard all samples (playback) or generate zero samples (capture)
oss
    Open Sound System
pipewire
    PipeWire Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
default
    Default ALSA Output (currently PipeWire Media Server)
sysdefault:CARD=sndxmosvocalfus
    snd_xmos_vocalfusion_card, simple-card_codec_link snd-soc-dummy-dai-0
    Default Audio Device
sysdefault:CARD=Headphones
    bcm2835 Headphones, bcm2835 Headphones
    Default Audio Device
**** List of PLAYBACK Hardware Devices ****
card 0: sndxmosvocalfus [snd_xmos_vocalfusion_card], device 0: simple-card_codec_link snd-soc-dummy-dai-0 [simple-card_codec_link snd-soc-dummy-dai-0]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 3: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 7/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7


ovos@OpenVoiceOS-840dca:~ $ arecord -lL
null
    Discard all samples (playback) or generate zero samples (capture)
oss
    Open Sound System
pipewire
    PipeWire Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
default
    Default ALSA Output (currently PipeWire Media Server)
sysdefault:CARD=sndxmosvocalfus
    snd_xmos_vocalfusion_card, simple-card_codec_link snd-soc-dummy-dai-0
    Default Audio Device
**** List of CAPTURE Hardware Devices ****
card 0: sndxmosvocalfus [snd_xmos_vocalfusion_card], device 0: simple-card_codec_link snd-soc-dummy-dai-0 [simple-card_codec_link snd-soc-dummy-dai-0]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
  ```

It is running this build with your piod changes and included the gpiod tools to it to be able to use the ```gpiodinfo``` command for debbuging.

https://github.com/OpenVoiceOS/ovos-buildroot/tree/fix/gpio

j1nx commented:
Forgot to mention

```
ovos@OpenVoiceOS-840dca:~ $ uname -a
Linux OpenVoiceOS-840dca 6.6.14-rt21-ovos-buildroot #1 SMP PREEMPT_RT Fri Feb  9 13:31:11 UTC 2024 aarch64 GNU/Linux

```
